src/stpipe/library.py

```diff
@@ -0,0 +1,850 @@
+            insensitive) a value in this list will be available
+            through the library.
+
+        asn_n_members : int, optional
+            Number of association members to include in the library.
+            This filtering will occur after "asn_exptypes" is applied.
+            By default all members will be included.
+
+        on_disk : bool, optional, default=False
+            When enabled, keep the models in the library "on disk",
+            writing out temporary files when the models are "shelved".
+            This option is only compatible with models loaded from
+            an association (not a list of models).
+
+        temp_directory : str or Path, optional
+            The temporary directory in which to store models when
+            "on_disk" is enabled. By default a
+            ``tempfile.TemporaryDirectory`` will be created in the
+            working directory.
+
+        **datamodels_open_kwargs : dict
+            Keyword arguments that will provided to ``_datamodels_open``
+            when models are opened.
+        """
+        self._on_disk = on_disk
+        self._open = False
+        self._ledger = _Ledger()
+        self._loaded_models = {}
+
+        self._datamodels_open_kwargs = datamodels_open_kwargs
+
+        if on_disk:
+            if temp_directory is None:
+                self._temp_dir = tempfile.TemporaryDirectory(dir="")
+                self._temp_path = Path(self._temp_dir.name)
+            else:
+                self._temp_path = Path(temp_directory)
+            self._temp_filenames = {}
+
+        if asn_exptypes is not None:
+            # if a single string, treat this as a single item list
+            if isinstance(asn_exptypes, str):
+                asn_exptypes = [asn_exptypes]
+            # convert these to lower case to allow case insensitive matching below
+            asn_exptypes = [exptype.lower() for exptype in asn_exptypes]
+
+        if isinstance(init, (str, Path)):
+            # init is an association filename (or path)
+            asn_path = os.path.abspath(os.path.expanduser(os.path.expandvars(init)))
+            self._asn_dir = os.path.dirname(asn_path)
+
+            # load association
+            asn_data = self._load_asn(asn_path)
+        elif isinstance(init, MutableMapping):
+            # we will modify the asn below so do a deep copy
+            asn_data = copy.deepcopy(init)
+            self._asn_dir = os.path.abspath(".")
+        elif isinstance(init, self.__class__):
+            raise ValueError(f"Invalid init {init}")
+        elif isinstance(init, Iterable):  # assume a list of models
+            if on_disk:
+                raise ValueError("on_disk cannot be used for a list of models")
+
+            # init is a list of models, this type of input ignores much of
+            # the benefits of this class but is common in test code.
+
+            # make a fake asn from the models
```

Comment:
A utility that generates an association from a list of models? I'm not aware of anything that quite "fits the bill". I'm not sure if stpipe is ready to house common association code. We might even need to make this association generation something the subclass has to define since there are some differences between jwst and roman association files (and the roman format as far as I'm aware is not finalized).

There is [asn_from_list](https://github.com/spacetelescope/jwst/blob/bc255f4cffa7a887ab3d1eed7ebfe82b856a4c67/jwst/associations/asn_from_list.py#L11) in jwst but that doesn't accept models. Although it is used in a [few](https://github.com/spacetelescope/jwst/blob/bc255f4cffa7a887ab3d1eed7ebfe82b856a4c67/jwst/wfs_combine/tests/test_wfs_combine.py#L77) [tests](https://github.com/spacetelescope/jwst/blob/bc255f4cffa7a887ab3d1eed7ebfe82b856a4c67/jwst/skymatch/tests/test_skymatch.py#L404) (these are just a few examples) to generate fake associations.